extension/visualizer/visualizer-extension.cpp

```diff
@@ -0,0 +1,126 @@
+#include "duckdb.hpp"
+#include "visualizer-extension.hpp"
+#include "duckdb/parser/parsed_data/create_pragma_function_info.hpp"
+#include "duckdb/catalog/catalog.hpp"
+#include "duckdb/common/fstream.hpp"
+#include "duckdb/main/query_profiler.hpp"
+#include "duckdb/main/client_context.hpp"
+#include "visualizer_constants.hpp"
+
+namespace duckdb {
+
+static string ToHTML(ClientContext &context, const string &first_json_path, const string &second_json_path) {
+	std::stringstream ss;
+	ss << "<!DOCTYPE <html>\n";
+	ss << "<style>\n";
+	ss << css;
+	ss << "</style>\n";
+	ss << "<head>\n";
+	ss << "\t<meta charset=\"utf-8\">\n";
+	ss << "\t<title>Query Profile Graph for Query</title>\n";
+	ss << "</head>\n";
+	ss << "<body>\n";
+	ss << "<script>";
+	ss << d3;
+	ss << "</script>\n";
+	ss << "<script> var data = ";
+	// If no json_file is given, read from query profiler
+	auto &prevProfilers = context.query_profiler_history->GetPrevProfilers();
+	if (first_json_path.empty() && !prevProfilers.empty()) {
+		ss << prevProfilers.back().second->ToJSON();
+	}
+	if (!first_json_path.empty()) {
+		ifstream data_json(first_json_path);
+		ss << data_json.rdbuf();
+	}
+	ss << "</script>\n";
+	ss << "<script> var secondData = ";
+	// If no json_file is given, make second json_data null
+	if (second_json_path.empty()) {
+		ss << "null;";
+	} else {
+		ifstream data_json(second_json_path);
+		ss << data_json.rdbuf();
+	}
+	ss << "</script>\n";
+	ss << "\n";
+	ss << "<script>";
+	ss << script;
+	ss << "</script>\n";
+	ss << "</body>\n";
+	ss << "</html>\n";
+	return ss.str();
+}
+
+static void WriteToFile(string &path, string info) {
+	ofstream out(path);
```

Comment:
What happens if the ofstream fails to write the file?

Reply:
If it fails then the HTML file will be corrupted. this piece of code is exactly like [link](https://github.com/duckdb/duckdb/blob/b04d8fe5d7e237eb60d91591895d3f90f6070793/src/main/query_profiler.cpp#L410)

Reply:
Could you fix it in both places and throw an exception in case of failure? Ideally errors should be caught early, rather than leading to corrupt output.